cmd/merge-walk-pool.go

```diff
@@ -26,13 +26,118 @@ const (
 	globalMergeLookupTimeout = time.Minute * 1 // 1 minutes.
 )
 
+// mergeWalkVersions - represents the go routine that does the merge walk versions.
+type mergeWalkVersions struct {
+	entryChs   []FileInfoVersionsCh
+	endWalkCh  chan struct{}   // To signal when mergeWalk go-routine should end.
+	endTimerCh chan<- struct{} // To signal when timer go-routine should end.
+}
+
 // mergeWalk - represents the go routine that does the merge walk.
 type mergeWalk struct {
 	entryChs   []FileInfoCh
 	endWalkCh  chan struct{}   // To signal when mergeWalk go-routine should end.
 	endTimerCh chan<- struct{} // To signal when timer go-routine should end.
 }
 
+// MergeWalkVersionsPool - pool of mergeWalk go routines.
+// A mergeWalk is added to the pool by Set() and removed either by
+// doing a Release() or if the concerned timer goes off.
+// mergeWalkPool's purpose is to maintain active mergeWalk go-routines in a map so that
+// it can be looked up across related list calls.
+type MergeWalkVersionsPool struct {
+	sync.Mutex
+	pool    map[listParams][]mergeWalkVersions
+	timeOut time.Duration
+}
+
+// NewMergeWalkVersionsPool - initialize new tree walk pool for versions.
+func NewMergeWalkVersionsPool(timeout time.Duration) *MergeWalkVersionsPool {
+	tPool := &MergeWalkVersionsPool{
+		pool:    make(map[listParams][]mergeWalkVersions),
+		timeOut: timeout,
+	}
+	return tPool
+}
+
+// Release - similar to mergeWalkPool.Release but for versions.
+func (t *MergeWalkVersionsPool) Release(params listParams) ([]FileInfoVersionsCh, chan struct{}) {
+	t.Lock()
+	defer t.Unlock()
+	walks, ok := t.pool[params] // Pick the valid walks.
+	if ok {
```

Comment:
```suggestion
	if !ok || len(walks) == 0 {
		return nil, nil
	}
```
Keep to the left. Technically we only need to check the length, since walks is nil if not found.